mm2src/coins/eth.rs

```diff
@@ -6130,52 +6354,165 @@ impl From<Web3RpcError> for EthGasDetailsErr {
         None => {
+            // If WithdrawFee not set use legacy gas price (?)
             let gas_price = eth_coin.get_gas_price().compat().await?;
-            // covering edge case by deducting the standard transfer fee when we want to max withdraw ETH
-            let eth_value_for_estimate = if fungible_max && eth_coin.coin_type == EthCoinType::Eth {
-                eth_value - gas_price * U256::from(21000)
-            } else {
-                eth_value
-            };
-            let estimate_gas_req = CallRequest {
-                value: Some(eth_value_for_estimate),
-                data: Some(data),
-                from: Some(eth_coin.my_address),
-                to: Some(call_addr),
-                gas: None,
-                // gas price must be supplied because some smart contracts base their
-                // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
-                gas_price: Some(gas_price),
-                ..CallRequest::default()
-            };
-            // TODO Note if the wallet's balance is insufficient to withdraw, then `estimate_gas` may fail with the `Exception` error.
-            // TODO Ideally we should determine the case when we have the insufficient balance and return `WithdrawError::NotSufficientBalance`.
-            let gas_limit = eth_coin.estimate_gas_wrapper(estimate_gas_req).compat().await?;
-            Ok((gas_limit, gas_price))
+            // go to gas estimate code
+            PayForGasOption::Legacy(LegacyGasPrice { gas_price })
+        },
+    };
+
+    // covering edge case by deducting the standard transfer fee when we want to max withdraw ETH
+    let eth_value_for_estimate = if fungible_max && eth_coin.coin_type == EthCoinType::Eth {
+        eth_value - calc_total_fee(U256::from(21000), &pay_for_gas_option)?
+    } else {
+        eth_value
+    };
+    let estimate_gas_req = CallRequest {
+        value: Some(eth_value_for_estimate),
+        data: Some(data),
+        from: Some(eth_coin.my_address),
+        to: Some(call_addr),
+        gas: None,
+        // gas price must be supplied because some smart contracts base their
+        // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
+        gas_price: if let PayForGasOption::Legacy(LegacyGasPrice { gas_price }) = pay_for_gas_option {
+            Some(gas_price)
+        } else {
+            None
+        },
+        max_priority_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas {
+            max_priority_fee_per_gas,
+            ..
+        }) = pay_for_gas_option
+        {
+            Some(max_priority_fee_per_gas)
+        } else {
+            None
+        },
+        max_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas { max_fee_per_gas, .. }) = pay_for_gas_option
+        {
+            Some(max_fee_per_gas)
+        } else {
+            None
+        },
+        ..CallRequest::default()
+    };
```

Comment:
please move fields creation before building CallRequest structure

```rust
    // gas price must be supplied because some smart contracts base their
    // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
    let gas_price = match pay_for_gas_option {
        PayForGasOption::Legacy(LegacyGasPrice { gas_price }) => Some(gas_price),
        _ => None,
    };
    let max_priority_fee_per_gas = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas {
            max_priority_fee_per_gas,
            ..
        }) => Some(max_priority_fee_per_gas),
        _ => None,
    };
    let max_fee_per_gas = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas { max_fee_per_gas, .. }) => Some(max_fee_per_gas),
        _ => None,
    };
    let estimate_gas_req = CallRequest {
        value: Some(eth_value_for_estimate),
        data: Some(data),
        from: Some(eth_coin.my_address),
        to: Some(call_addr),
        gas: None,
        gas_price,
        max_priority_fee_per_gas,
        max_fee_per_gas,
        ..CallRequest::default()
    };
 ```
 Its hard to read when there is incorporation of if let expression (or match pattern) directly within the creation of a structure.

Reply:
done in [8230a89](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/8230a8903b29f0f2492ccc4f101e4a72c09e09ab)
(yes this looks better, thanks)

Reply:
I would like to suggest to polish code a bit more. here I see that match pattern is more preferable

https://github.com/KomodoPlatform/komodo-defi-framework/blob/66f8ff76558e8330d08c6694d9e4efc6dbc3b7a3/mm2src/coins/eth.rs#L6434-L6448

```rust
    // Legacy transactions use a single gas price, while EIP-1559 transactions specify fees differently.
    let gas_price = match pay_for_gas_option {
        PayForGasOption::Legacy(LegacyGasPrice { gas_price }) => Some(gas_price),
        _ => None,
    };

    // Extract priority and max fees for EIP-1559 transactions, if applicable.
    let (max_priority_fee_per_gas, max_fee_per_gas) = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas {
            max_priority_fee_per_gas,
            max_fee_per_gas,
        }) => (Some(max_priority_fee_per_gas), Some(max_fee_per_gas)),
        _ => (None, None),
    };

    let estimate_gas_req = CallRequest {
        value: Some(eth_value_for_estimate),
        data: Some(data),
        from: Some(eth_coin.my_address),
        to: Some(call_addr),
        gas: None,
        // gas price must be supplied because some smart contracts base their
        // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
        gas_price,
        max_priority_fee_per_gas,
        max_fee_per_gas,
        ..CallRequest::default()
    };
```